From 22218d666d8b9009dce5210d16289eb6bf34bae0 Mon Sep 17 00:00:00 2001
From: Han Xu <b45815@freescale.com>
Date: Thu, 28 May 2015 16:49:18 -0500
Subject: [PATCH 0230/5242] MLK-10985: mtd: nand: support NAND on i.MX6UL

commit  4ad1eed3d2f1b30ece6bf34425e7fcb2c9acaad3 from
https://source.codeaurora.org/external/imx/linux-imx.git

support i.MX6UL GPMI NAND driver and removed the unecessary clock
per1_bch.

Signed-off-by: Han Xu <b45815@freescale.com>

During 4.14 rebase fixed for upstream moving clocks to gpmi_devdata

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c |   13 ++++++++++++-
 drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h |    5 ++++-
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
index 090ceea..a168b7b 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
@@ -108,7 +108,7 @@ static int gpmi_ooblayout_free(struct mtd_info *mtd, int section,
 };
 
 static const char * const gpmi_clks_for_mx6[] = {
-	"gpmi_io", "gpmi_apb", "gpmi_bch", "gpmi_bch_apb", "per1_bch",
+	"gpmi_io", "gpmi_apb", "gpmi_bch", "gpmi_bch_apb",
 };
 
 static const struct gpmi_devdata gpmi_devdata_imx6q = {
@@ -133,6 +133,14 @@ static int gpmi_ooblayout_free(struct mtd_info *mtd, int section,
 	.clks_count = ARRAY_SIZE(gpmi_clks_for_mx6),
 };
 
+static const struct gpmi_devdata gpmi_devdata_imx6ul = {
+	.type = IS_MX6UL,
+	.bch_max_ecc_strength = 40,
+	.max_chain_delay = 12,
+	.clks = gpmi_clks_for_mx6,
+	.clks_count = ARRAY_SIZE(gpmi_clks_for_mx6),
+};
+
 static const char * const gpmi_clks_for_mx7d[] = {
 	"gpmi_io", "gpmi_bch_apb",
 };
@@ -2004,6 +2012,9 @@ static int gpmi_nand_init(struct gpmi_nand_data *this)
 		.compatible = "fsl,imx6sx-gpmi-nand",
 		.data = &gpmi_devdata_imx6sx,
 	}, {
+		.compatible = "fsl,imx6ul-gpmi-nand",
+		.data = (void *)&gpmi_devdata_imx6ul,
+	}, {
 		.compatible = "fsl,imx7d-gpmi-nand",
 		.data = &gpmi_devdata_imx7d,
 	}, {}
diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
index 692d7b5..87dbc3a 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
@@ -22,7 +22,7 @@
 #include <linux/dma-mapping.h>
 #include <linux/dmaengine.h>
 
-#define GPMI_CLK_MAX 5 /* MX6Q needs five clocks */
+#define GPMI_CLK_MAX 4 /* MX6Q needs four clocks */
 struct resources {
 	void __iomem  *gpmi_regs;
 	void __iomem  *bch_regs;
@@ -84,6 +84,7 @@ enum gpmi_type {
 	IS_MX6QP,
 	IS_MX6SX,
 	IS_MX7D,
+	IS_MX6UL
 };
 
 struct gpmi_devdata {
@@ -215,7 +216,9 @@ void gpmi_copy_bits(u8 *dst, size_t dst_bit_off,
 #define GPMI_IS_MX6QP(x)	((x)->devdata->type == IS_MX6QP)
 #define GPMI_IS_MX6SX(x)	((x)->devdata->type == IS_MX6SX)
 #define GPMI_IS_MX7D(x)		((x)->devdata->type == IS_MX7D)
+#define GPMI_IS_MX6UL(x)	((x)->devdata->type == IS_MX6UL)
 
 #define GPMI_IS_MX6(x)		(GPMI_IS_MX6Q(x) || GPMI_IS_MX6QP(x) || GPMI_IS_MX6SX(x) || \
+                                 GPMI_IS_MX6UL(x) || \
 				 GPMI_IS_MX7D(x))
 #endif
-- 
1.7.9.5

